File: python/2015/14/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def step(a, s):
    ret = 0
    (speed, time, rest) = a
    while s > time+rest:
        ret += speed*time
        s -= time+rest
    if s >= time:
        ret += speed*time
    else:
        ret += speed*s
    return ret

def iter(A, n):
    l = [0] * len(A)
    for s in range(n):
        for i,a in enumerate(A):
            l[i] = step(a, s+1)
    return max(l)

def point(A, n):
    l = [0] * len(A)
    p = [0] * len(A)
    for s in range(n):
        for i,a in enumerate(A):
            l[i] = step(a, s+1)
        m = 0
        for i in range(len(A)):
            if m < l[i]:
                m = l[i]
        for i in range(len(A)):
            if l[i] == m:
                p[i] += 1
    return max(p)

def read_input():
    X = u.read_match(file,".* ([0-9]+) .* ([0-9]+) .* ([0-9]+) .*")
    return [(int(x[0]), int(x[1]), int(x[2])) for x in X]

def part1():
    A = read_input()
    k = iter(A, 2503)
    print("Part 1:",k)

def part2():
    A = read_input()
    k = point(A, 2503)
    print("Part 2:",k)

part1()
part2()

File: c/2015/14/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

int64_t step(uint8_t deer[], int64_t time) {
    int64_t ret = 0;
    int64_t t = time;
    while (t > deer[1]+deer[2]) {
        ret += deer[0]*deer[1];
        t -= deer[1]+deer[2];
    }
    if (t >= deer[1]) {
        ret += deer[0]*deer[1];
    } else {
        ret += deer[0]*t;
    }
    return ret;
}

int64_t iter(uint8_t deer[][3], int16_t length, int16_t time) {
    int16_t l[length];
    for (int64_t i = 0; i < time; i++) {
        for (int16_t j = 0; j < length; j++) {
            l[j] = step(deer[j], i+1);
        }
    }
    uint64_t max = 0;
    for (int16_t j = 0; j < length; j++) {
        if (max < l[j]) {
            max = l[j];
        }
    }
    return max;
}

int64_t point(uint8_t deer[][3], int16_t length, int16_t time) {
    int64_t l[length];
    int64_t p[length];
    for (int16_t j = 0; j < length; j++) {
        p[j] = 0;
    }
    for (int64_t i = 0; i < time; i++) {
        for (int16_t j = 0; j < length; j++) {
            l[j] = step(deer[j], i+1);
        }
        uint64_t max = 0;
        for (int16_t j = 0; j < length; j++) {
            if (max < l[j]) {
                max = l[j];
            }
        }
        for (int16_t j = 0; j < length; j++) {
            if (l[j] == max) {
                p[j]++;
            }
        }
    }
    uint64_t max = 0;
    for (int16_t j = 0; j < length; j++) {
        if (max < p[j]) {
            max = p[j];
        }
    }
    return max;
}

void part1() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ");
    uint8_t deers[9][3];
    for (int8_t i = 0; i < in->length; i++) {
        deers[i][0] = atoi(in->strings[i].strings[3]);
        deers[i][1] = atoi(in->strings[i].strings[6]);
        deers[i][2] = atoi(in->strings[i].strings[13]);
    }
    int64_t k = iter(deers, in->length, 2503);
    printf("Part 1: %ld\n", k);
    freeArrMulString(in);
}

void part2() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ");
    uint8_t deers[9][3];
    for (int8_t i = 0; i < in->length; i++) {
        deers[i][0] = atoi(in->strings[i].strings[3]);
        deers[i][1] = atoi(in->strings[i].strings[6]);
        deers[i][2] = atoi(in->strings[i].strings[13]);
    }
    int64_t k = point(deers, in->length, 2503);
    printf("Part 2: %ld\n", k);
    freeArrMulString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
